cpp/test/Slice/errorDetection/TypeAlias.ice

```diff
@@ -0,0 +1,80 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+module Test
+{
+    sequence<float> floatSeq;
+
+    struct Custom
+    {
+        int i;
+    }
+
+    class C0 {}
+    interface I0 {}
+    exception E0 {}
+
+    module Nested1
+    {
+        class C1;
+    }
+
+    module Nested2
+    {
+        class C2;
+    }
+
+    // Basic type aliases
+    using intAlias = int;               //ok
+    using stringAlias = string;         //ok
+    using floatSeqAlias = floatSeq;     //ok
+    using customAlias = Custom;         //ok
+    using unknownAlias = foo;           //can't resolve type
+    using optintAlias = int?;           //can't typealias optional types
+    using missingType;                  //missing underlying type
+    using aliasName = int               //missing semicolon
+
+    using C0Alias = C0;                      //ok
+    using C1Alias = C1;                      //can't resolve type
+    using scopedC0Alias = Test::C0;          //ok
+    using scopedC1Alias = Test::Nested1::C1; //ok
+
+    // Type resolution of type-aliases
+    module Nested2
+    {
+        using C0Alias = C0;                      //ok
+        using C1Alias = C1;                      //can't resolve type
+        using C2Alias = C2;                      //ok
+        using scopedC0Alias = Test::C0;          //ok
+        using scopedC1Alias = Test::Nested1::C1; //ok
+        using scopedC2Alias = Test::Nested2::C2; //ok
+    }
+
+    // Inheritance from type-aliases
+    // Note, all metadata is allowed, but non-type metadata will error out when the alias is used.
+    using C0Alias = C0;                            //redefinition error
+    using C0AliasMetadata = [protected] C0;        //ok
+    class DerivedClass1 : C0Alias {}               //ok
+    class DerivedClass2 : C0AliasMetadata {}       //illegal metadata
+
+    using I0Alias = I0;                            //ok
+    using I0AliasMetadata = [amd] I0;              //ok
```

Comment:
I would expect this to be an error

Reply:
The way I've written type-aliases is that any metadata can be applied in any type-alias, and errors are only reported when the type-alias is actually used. So if you used `I0AliasMetadata` somewhere, it would raise an error for invalid metadata.

This seemed more flexible, since metadata validity can be context dependent.
Plus, right now, all our metadata validation is performed _after_ the parsing stage (usually by metadata visitors), when type-aliases are already resolved. It seems simpler just to use our existent metadata validation, instead of adding another validation step in the parsing stage.

Reply:
maybe we can delegate the validation to the type being aliases and the validation should decide if the metadata is valid for the type and for the type alias, I think [amd],[prtected] only make sense to accept it as metadata for the type but not for the alias. Anyway this is something we can rework later.

Reply:
Yes, I agree that `[amd]` and `[protected]` are totally bogus here! I don't have any issue with moving the validation onto the alias, but it'll probably have to wait until after I get back to my metadata-related changes... Right now our metadata validation just isn't very powerful or flexible.

Reply:
Same as above, I'll rework this once we have stronger metadata logic so we can filter type and non-type metadata.